060-kafka-roller-kraft.md

```diff
@@ -0,0 +1,159 @@
+- Count each controller node that is caught up (`numOfCaughtUpControllers`).
+- Can restart if: `numOfCaughtUpControllers >= ceil((double) (totalNumOfControllers + 1) / 2)`.
+
+> NOTE: Until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented, KafkaRoller cannot create an admin connection to the controller directly to describe its configuration or the quorum state. Therefore KafkaRoller checks the desired configurations to get the value of `controller.quorum.fetch.timeout.ms` and creates admin client connection to the brokers for the quorum check. If KafkaRoller cannot connect to any of the brokers after the maximum attempts to retry, the controllers will be marked as failed to reconcile because of not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. In this scenario, the reconcilation will be completed with failure and reported to the operator. The controllers quorum check will be retried in the next round of reconcilation.
+
+
+#### Separate brokers and controllers
+For controller-only:
+ 1. Does not restart a controller performing log recovery
+ 2. Restarts a controller if the pod is in stuck state
+ 3. Attempts to connect an admin client to any of the brokers, if cannot connect, return `UnforceableProblem` which results in delay and retry for the pod until the maximum attempts is reached. 
+When maximum attempts is reached the KafkaRoller will move on to restart broker pods incase this resolves connection issues but will still ultimately mark the reconciliation as failed.
+ 4. Restarts a controller if the controller configuration has changed
+ 5. Does not restart a controller if doing so would take the number of caught-up controllers (inc leader) to less than majority of the quorum
+
+For broker-only:
+ 1. Does not restart a broker performing log recovery
+ 2. Restarts a broker if the pod is in stuck state
+ 3. Attempts to connect an admin client to a broker and if it can't connect restarts the broker
+ > **NEW** Currently when pod is stuck and out of date, it gets restarted because admin client connection fails. With this change, we will not attempt admin client connection to the broker, if the pod is stuck. Instead the pod will be restarted without making the admin client connection. Therefore the resulting action that restarts the pod that is stuck stays the same for zookeeper mode. This resolves the issue mentioned at the beginning of the proposal that leaves some pods stuck in pending state.
```

Comment:
```suggestion
 > **NEW** Currently when pod is stuck and out of date, it is restarted because admin client connection fails. With this change, we will not attempt admin client connection to the broker if the pod is stuck. Instead the pod will be restarted without making the admin client connection. Therefore, the resulting action that restarts the pod that is stuck stays the same for ZooKeeper mode. This resolves the issue mentioned at the beginning of the proposal that leaves some pods stuck in pending state.
```